File: DOZO-Server/src/lib/terminal-id.js
const TERMINAL_ID_RE = /^[A-Za-z0-9_-]{8,64}$/;

/** True when a caller-supplied terminal_id is URL-safe and 8-64 chars. */
export function isValidTerminalId(value) {
  return typeof value === 'string' && TERMINAL_ID_RE.test(value);
}

/**
 * Derive a stable terminal_id from a device serial when the client does not
 * supply one: uppercase, strip everything outside [A-Z0-9], truncate to 32,
 * and zero-pad to a minimum length of 8.
 */
export function deriveTerminalId(deviceSerial) {
  const normalized = String(deviceSerial).toUpperCase().replace(/[^A-Z0-9]/g, '');
  const truncated = normalized.slice(0, 32);
  return truncated.length >= 8 ? truncated : truncated.padEnd(8, '0');
}

File: DOZO-App/app/src/main/java/com/example/dozo/TerminalId.kt
package com.example.dozo

object TerminalId {

    private const val MIN_LENGTH = 8
    private const val MAX_LENGTH = 16
    private val NON_ALPHANUMERIC = Regex("[^A-Za-z0-9]")

    fun derive(seed: String): String {
        val normalized = seed.uppercase().replace(NON_ALPHANUMERIC, "")
        return normalized.take(MAX_LENGTH).padEnd(MIN_LENGTH, '0')
    }
}
